yt_dlp/postprocessor/ffmpeg.py

```diff
@@ -1169,3 +1178,226 @@ def run(self, info):
             'ext': ie_copy['ext'],
         }]
         return files_to_delete, info
+
+
+class FFmpegProgressTracker:
+
+    def __init__(self, info_dict, ffmpeg_args, hook_progress, ydl=None):
+        self.ffmpeg_proc = Popen(ffmpeg_args, env=environ.copy())
+        self.ydl = ydl
+        self._info_dict = info_dict
+        self._ffmpeg_args = ffmpeg_args
+        self._hook_progress = hook_progress
+        self._out_queue = Queue()
+        self._err_queue = Queue()
+        self._stdout_buffer = ''
+        self._stderr_buffer = ''
+        self._progress_pattern = re.compile(
+            r'(frame=\s*(?P<frame>\S+)\n'
+            r'fps=\s*(?P<fps>\S+)\n'
+            r'stream_\d+_\d+_q=\s*(?P<stream_d_d_q>\S+)\n)?'
+            r'bitrate=\s*(?P<bitrate>\S+)\n'
+            r'total_size=\s*(?P<total_size>\S+)\n'
+            r'out_time_us=\s*(?P<out_time_us>\S+)\n'
+            r'out_time_ms=\s*(?P<out_time_ms>\S+)\n'
+            r'out_time=\s*(?P<out_time>\S+)\n'
+            r'dup_frames=\s*(?P<dup_frames>\S+)\n'
+            r'drop_frames=\s*(?P<drop_frames>\S+)\n'
+            r'speed=\s*(?P<speed>\S+)\n'
+            r'progress=\s*(?P<progress>\S+)'
+        )
+
+    def run_ffmpeg_subprocess(self):
+        if self._info_dict:
+            return self._track_ffmpeg_progress()
+        return self._run_ffmpeg_without_progress_tracking()
+
+    def _run_ffmpeg_without_progress_tracking(self):
+        """Simply run ffmpeg and only care about the last stderr, stdout and the retcode"""
+        stdout, stderr = self.ffmpeg_proc.communicate_or_kill()
+        retcode = self.ffmpeg_proc.returncode
+        return stdout, stderr, retcode
+
+    def _track_ffmpeg_progress(self):
+        """ Track ffmpeg progress in a non blocking way using queues"""
+        self._start_time = time.time()
+        # args needed to track ffmpeg progress from stdout
+        self._ffmpeg_args.extend(['-progress', 'pipe:1'])
+        self._duration_to_track, self._total_duration = self._compute_duration_to_track()
+        self._total_filesize = self._compute_total_filesize(self._duration_to_track, self._total_duration)
+        self._status = {
+            'filename': self._info_dict.get('_filename'),
```

Comment:
1. The file being modified need not the video file. This should be inferred from args
1. For postprocessor hook, the API is not currently defined. To account for multiple files being modified, 'filenames': [...] may be more appropriate
1. Same issue exists for filesize and duration, but worse. Eg: Convert subtitles, convert thumbnails, split chapters, concat video